### ISSUE TITLE
How to use database transactions with poem-openapi?

### ISSUE THREAD

**Issue by TheCataliasTNT2k** [poem-web/poem#479]
I created this example based on https://github.com/poem-web/poem/blob/master/examples/openapi/log-with-operation-id/src/main.rs:
```rust
use time::OffsetDateTime;
use poem::{listener::TcpListener, Endpoint, EndpointExt, Route, Server, Error};
use poem::http::StatusCode;
use poem::middleware::Cors;
use poem::web::Data;
use poem_openapi::{param::Query, payload::PlainText, OpenApi, OpenApiService};

struct Api;

#[OpenApi]
impl Api {
    #[oai(path = "/hello", method = "get", operation_id = "index-get")]
    async fn index(&self, name: Query<Option<String>>, data: Data<&AppState>, t: Data<&OffsetDateTime>) -> Result<PlainText<String>, Error> {
        println!("{}, {}", data.config, t.to_string());
        if t.second() % 2 == 0 {
            Ok(
                match name.0 {
                    Some(name) => PlainText(format!("hello, {}!", name)),
                    None => PlainText("hello!".to_string()),
                }
            )
        } else {
            Err(Error::from_status(StatusCode::FORBIDDEN))
        }
    }
}


#[derive(Debug, Clone)]
struct AppState {
    config: String,
}

#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    let api_service =
        OpenApiService::new(Api, "Hello World", "1.0").server("http://localhost:3000/api");
    let ui = api_service.swagger_ui();
    let state = AppState {
        config: "Test".to_string(),
    };
    let app = Route::new()
        .nest("/api", api_service)
        .nest("/", ui)
        .around(|ep, req| async move {
            let now = OffsetDateTime::now_utc();
            let resp = ep.clone().data(now).get_response(req).await;

            if resp.status().is_success() {
                println!("No problem")
            } else {
                println!("Problem")
            }

            Ok(resp)
        })
        .data(state)
        .with(Cors::new());

    Server::new(TcpListener::bind("127.0.0.1:3000"))
        .run(app)
        .await
}

```

The idea is to

1. create a database transaction for each request
2. provide it as parameter
3. do stuff with it
4. commit or roll it back, depending on the status code of the response

I updated the example in the sea orm repo, here is the PR: https://github.com/SeaQL/sea-orm/pull/1385
I am using SeaOrm for my database stuff, but I did not include it in this example to keep it as simple as possible.
I do not know, if there is a better way than shown above, but I have a few problems and possible with this:

- I need to put the database connection into `AppState` even though no endpoint uses it (I found no other way to get it into the async closure within `around`)
- The `.data()` needs to be after the `.around()`, otherwise the db connection can not be found
- I could not find a way to supply the endpoints with a non-cloneable transaction. I need the object to provide it for the endpoint, and I need it later for `commit` or `rollback`.
So if the transaction gets cloned within the endpoint, you might run into issues.
Possible solutions: Providing only a (mutable) reference to the endpoint.
Or: Getting the object back from poem after the endpoint finishes.
But Poem does not support either, as far as I see.

Maybe you find better ways and can improve the code?

**Comment by sunli829** [comment]
If the extensions in the request is moved to the response after the request is processed, and you can get it, so that you can take `DatabaseTransaction` from the response, there is no need for `Arc`, what do you think?

**Comment by TheCataliasTNT2k** [comment]
Yes, but is is not moved there, that is one of the problems :wink: 
I already tried that...

**Comment by tgrushka** [comment]
I wanted this too with sqlx. I tried to create middleware to wrap a request in a `sqlx::Transaction`, but `sqlx::Transaction` is not `Clone`, which is required to store it using `Request::set_data(...)`. (I was able to get code to compile by wrapping it in an `Arc<Mutex<Option<Transaction<'_, Postgres>>>>`, but who knows if it works -- it looks horrible and I won't even try to run it.)

### At least 3 other Rust web frameworks have the ability to allow transaction middleware support for sqlx (via external crates):

* axum-sqlx-tx
* actix-sqlx-tx
* tide-sqlx

Why not poem?
